variant/effect.go

```diff
@@ -0,0 +1,276 @@
+package variant
+
+import (
+	"github.com/vertgenlab/gonomics/dna"
+)
+
+// CodingChange describes a change of a protein sequence caused by
+// a mutation affecting the coding sequence of the corresponding gene.
+// CodingPos and ProteinPos refer to the zero-based start position of
+// the change. The CodingPos always refers to the position changed in
+// origin DNA sequence, however ProteinPos is not always equal to
+// CodingPos / 3. If a change would result in a RemovedAa[0] == AddedAa[0]
+// the first amino acid from each slice is removed and ProteinPos += 1.
+//
+// RemovedAa records all amino acids that are removed from the protein
+// sequence. AddedAa records all amino acids that are added to the
+// protein sequence. Substitution of one amino acid for another is
+// reported as the original amino acid being removed, and the new amino
+// acid being added. dna.Stop is included in RemovedAa and AddedAa.
+type CodingChange struct {
+	CodingPos  int
+	ProteinPos int
+	RemovedAa  []dna.AminoAcid
+	AddedAa    []dna.AminoAcid
+	Type       EffectType
+}
+
+// EffectType describes the overall effect of a CodingChange.
+// Possible EffectTypes are described below.
+type EffectType byte
+
+const (
+	Silent           EffectType = 6 // mutation changes codon, but still codes for the same amino acid
+	Frameshift       EffectType = 5 // shifts reading frame
+	Nonsense         EffectType = 4 // premature termination
+	InFrameInsertion EffectType = 3 // insertion where insertedBases % 3 == 0
+	InFrameDeletion  EffectType = 2 // deletion where deletedBases % 3 == 0
+	Missense         EffectType = 1 // mutation changes coded amino acid
+)
+
+// Effector is satisfied by types that implement the Effect method, which returns
+// a CodingChange describing how changes prescribed by the receiver type alter the
+// resulting amino acid sequence corresponding to the input []dna.Base. The input
```

Comment:
Why not have the input []dna.Base be the whole chromosome?  While it might seem wasteful, that way we only need to store the array once and have different slices use the same array.  I don't quite have the big picture of if I have a bunch of genome-wide mutations and a file listing my genes, how I can then start using these functions.  Would I first truncate my deletions to only cover the protein coding seq?

Reply:
You could, but it complicates things greatly. You would need to input CDS starts and ends and the program would become far more complicated. The vision here is to create a fairly simple system that we can build off of. The creation of the CDS sequence would be handled by higher level programs that use the variant infrastructure build here (e.g. the gene package). Having written the complicated version in other packages, I think the tradeoff of efficiency for simplicity is worth it here.

Reply:
To answer your second question, I plan to build things so you can take an input VCF, GTF, FASTA. The GTF/FASTA are converted to Gene structs, and the VCF is converted to variants/Effectors. Then you could make a 
```
func AnnotateVariant(g Gene, e Effector) {
      return e.Effect(g.CDS)
}
```